Refactor filestore API so that it uses internal TileDB library calls.

abigalekim opened the pull request:
I am currently refactoring `tiledb/c_api/tiledb_filestore.cc` so that it uses only internal API calls, as opposed to the C++ external API calls.

---
TYPE: IMPROVEMENT
DESC: Refactor filestore API so that it uses internal TileDB library calls.

robertbindar commented on tiledb/sm/c_api/tiledb_filestore.cc:
```diff
@@ -270,70 +318,93 @@ TILEDB_EXPORT int32_t tiledb_filestore_uri_import(
+  throw_if_not_ok(subarray.add_range(0, std::move(subarray_range)));
   query.set_subarray(subarray);
 
   auto tiledb_cloud_fix = [&](uint64_t start, uint64_t end, uint64_t nbytes) {
```
safe to remove `nbytes`, `end - start + 1` should be enough within the function.

robertbindar commented on tiledb/sm/c_api/tiledb_filestore.cc:
```diff
@@ -388,40 +472,56 @@ TILEDB_EXPORT int32_t tiledb_filestore_uri_export(
     // Cloud compatibility hack. Currently stored tiledb file arrays have a
     // TILEDB_UINT8 attribute. We should pass the right datatype here to
     // support reads from existing tiledb file arrays.
+    auto&& [st, schema] = array->get_array_schema();
```
preserve the comment block position

abigalekim replied:
Hmmm, I think it is preserved? I just have to get the schema before I extract the attribute type, so it seems like it is not.

robertbindar replied:
My bad, you're correct

robertbindar commented on tiledb/sm/c_api/tiledb_filestore.cc:
```diff
@@ -516,40 +635,63 @@ TILEDB_EXPORT int32_t tiledb_filestore_buffer_export(
-  if (file_size && *static_cast<const uint64_t*>(file_size) < offset + size) {
+      &file_size));
+  if (!file_size) {
+    auto st = Status_Error(
```
maybe just throw given this is all new code

robertbindar commented on tiledb/sm/c_api/tiledb_filestore.cc:
```diff
@@ -516,40 +635,63 @@ TILEDB_EXPORT int32_t tiledb_filestore_buffer_export(
+    save_error(ctx, st);
+    return TILEDB_ERR;
+  } else if (
+      file_size && *static_cast<const uint64_t*>(file_size) < offset + size) {
```
remove redundant `file_size` check

robertbindar commented on tiledb/sm/c_api/tiledb_filestore.cc:
```diff
@@ -560,27 +702,50 @@ TILEDB_EXPORT int32_t tiledb_filestore_size(
+
+  if (!file_size) {
+    auto st = Status_Error(
+        "The array metadata doesn't contain the " +
```
The existence of the metadata key is already checked above, no need to do it again here.

abigalekim replied:
but we should still null check the file_size attribute. I could throw a std::logic_error here and this is what I will do for now.

robertbindar replied:
We checked for the key existence prior to getting file_size. Any internal error in metadata that would render file_size=nullptr should be caught in the previous `throw_if_not_ok(array.get_metadata...`. I would say it's safe to assume file_size != nullptr here at this point.

robertbindar commented on tiledb/sm/c_api/tiledb_filestore.cc:
```diff
@@ -683,16 +848,12 @@ bool libmagic_file_is_compressed(void* data, uint64_t size) {
-  }
-  fb.close();
+    tiledb::sm::VFS& vfs, const char* uri, std::vector<char>& header) {
+  // TODO: I am very unsure if this is right...
```
looks fine

robertbindar commented on tiledb/sm/c_api/tiledb_filestore.cc:
```diff
@@ -270,70 +318,93 @@ TILEDB_EXPORT int32_t tiledb_filestore_uri_import(
+    return readlen;
+  };
+
+  // TODO: very unsure about this code.
```
safe to remove

robertbindar reviewed (approved):
Excellent work @abigalekim, thanks! I've left minor comments below. Also I double checked indices are right, please triple check. I advise you test manually that the cloud hack still works (just get rid of the `is_remote` check so that filestore writes ROW_MAJOR locally).

abigalekim commented on tiledb/sm/c_api/tiledb_filestore.cc:
```diff
@@ -62,10 +71,10 @@ std::pair<Status, optional<std::string>> libmagic_get_mime_encoding(
     void* data, uint64_t size);
 bool libmagic_file_is_compressed(void* data, uint64_t size);
 Status read_file_header(
-    const VFS& vfs, const char* uri, std::vector<char>& header);
```
I can't keep the const. VFS::open_file, VFS::read, and VFS::close_file are not marked const and I don't think they should be, since they internally change the state of the VFS.

abigalekim commented:
Hello! I addressed the most of the comments and think it works. However, I still don't know what you mean when you say  I advise you test manually that the cloud hack still works (just get rid of the is_remote check so that filestore writes ROW_MAJOR locally).

I see on line 292 (`  bool is_tiledb_uri = array->is_remote();`) this `is_tiledb_uri` variable is defined, and then used in 2 places. One to finalize the query and once to case on the cloud fix function. I deleted the casing on the cloud fix function and the cloud fix function definition itself, and it does work locally. Is this what you mean?

robertbindar commented:
Hey @abigalekim! The code places you mentioned above only get executed when the filestore array uri is `tiledb://` (a special case we added to make filestore work with tiledb cloud in the absence of global order writes serialization feature). There is no way to test this `tiledb://` hack currently other than manually using either:
- set up a rest server locally, write a small example, import a file, export back, diff with original.
- set in the filestore code `is_tiledb_uri = true` somewhere above the checks, this will make imports always use row_major (basically simulate an array is a tiledb:// array to hit those code places), write a small example that imports, exports, diffs a file.

abigalekim commented:
Hello, I have tested with is_tiledb_uri = true and the code passes--does this suffice for manual testing?

robertbindar reviewed (approved):
Sounds good to me.

ihnorton reviewed (changes requested):
A couple update requests:
- let's make the implementations throw consistently, right now there is a mix of status and throw
- there are a still a lot C-API isms (like `sanity_check` and `save_error`). Most of those should go away after addressing point (1), but please try to compile without `tiledb.h` and see if anything else is left
- remove use of `TILEDB_EXPORT` annotations

ihnorton commented:
cc @robertbindar. Also, it seems like there _might_ be a possible cmake change available to avoid linking this against the cpp_api (but maybe that was always implicit).

robertbindar commented on tiledb/sm/c_api/tiledb_filestore.cc:
```diff
@@ -31,238 +31,260 @@
     }
   }
 
   *array_schema = new (std::nothrow) tiledb_array_schema_t;
```
I think the new layer of CAPI catches bad allocs, `nothrow` doesn't make sense here anymore.

robertbindar commented on tiledb/sm/c_api/tiledb_filestore.cc:
```diff
@@ -31,238 +31,260 @@
   }
 
   *array_schema = new (std::nothrow) tiledb_array_schema_t;
   if (*array_schema == nullptr) {
```
This means this check and throw doesn't make sense either, we should just let `operator new` throw

robertbindar commented on tiledb/sm/c_api/tiledb_filestore.cc:
```diff
@@ -31,238 +31,260 @@
-    Domain domain(context);
-    domain.add_dimension(dim);
+    auto domain = make_shared<tiledb::sm::Domain>(HERE());
+    throw_if_not_ok(domain.get()->add_dimension(dim));
```
`.get()` doesn't seem necessary

robertbindar commented on tiledb/sm/c_api/tiledb_filestore.cc:
```diff
@@ -31,238 +31,260 @@
+          tiledb::sm::Compressor::ZSTD,
+          tiledb::sm::ZStd::default_level(),
+          tiledb::sm::Datatype::ANY));
+      attr.get()->set_filter_pipeline(filter);
```
`.get()` is unnecessary

robertbindar commented on tiledb/sm/c_api/tiledb_filestore.cc:
```diff
@@ -272,158 +294,194 @@ int32_t tiledb_filestore_uri_import(
-    query.set_data_buffer(
-        tiledb::sm::constants::filestore_attribute_name, buffer.data(), nbytes);
-    query.submit();
+  auto tiledb_cloud_fix = [&](uint64_t start, uint64_t end) {
```
Please create a followup task for switching filestore API to using global order writes natively for tiledb URIs.

KiterLuc replied:
https://app.shortcut.com/tiledb-inc/story/36254/filestore-api-should-use-global-order-writes

robertbindar commented on tiledb/sm/c_api/tiledb_filestore.cc:
```diff
@@ -272,158 +294,194 @@ int32_t tiledb_filestore_uri_import(
 
-  output.flush();
-  fb.close();
+  throw_if_not_ok(vfs.sync(tiledb::sm::URI(file_uri)));
```
not necessary, close_file does that already.

robertbindar reviewed (changes requested):
Minor things to change, please also create a followup that will trac enabling remote global order writes.

robertbindar reviewed (approved):
Thanks!
